About callback and added min build script at package.json

devmario opened the pull request:
#16 moved this PR
- Added callback chainable style 
- Callback timing issue fix
- Update readme

More commited
- Added min.js build script with uglify-js at devDependencies
```
npm install
npm build:min
```

SaulDoesCode commented:
perhaps in the .complete , .begin and .update functions return the anim object strait instead of this;
it will make binding and calling easier for those using it later.
also maybe use the type checker defined higher up.
```javascript
    anim.begin = function(callback) {
        if(!is.func(callback)) return anim;
        var s = this.settings;
        s.begin = callback;
        return anim;
    }

    anim.update = function(callback) {
        if(!is.func(callback)) return anim;
        var s = this.settings;
        s.update = callback;
        return anim;
    }

    anim.complete = function(callback) {
        if(!is.func(callback)) return anim;
        var s = this.settings;
        s.complete = callback;
        return anim;
    }

```

devmario commented:
Thx @SaulDoesCode.
I implemente just now.

SaulDoesCode commented:
do it in one line as it has the same outcome returning the anim object;
```javascript
        if(callback == undefined) return this;
        if(!is.func(callback)) return this;

        // becomes
       if(is.undef(callback) || !is.func(callback) return anim;
```

devmario commented:
Re thx @SaulDoesCode.
I implemente just now.

devmario commented on anime.js:
```diff
@@ -590,22 +590,33 @@
     }
 
     anim.update = function(callback) {
-        if(is.undef(callback) || !is.func(callback)) return this;
```
Is.func
```js
var is = (function() {
    Return {
      ...
      func:   function(a) { return typeof a === 'function' },
      ...
   }
}
```

SaulDoesCode replied:
that is what I have had in my es6 version and this new v1.1 version

SaulDoesCode replied:
Why do you undefine the callback, what if a former callback already exists?

devmario replied:
@SaulDoesCode 
For example
```js
anim.play()
.complete(function(anim) {
   anim.play().complete(); // This code is complete callback disable
});
```

SaulDoesCode replied:
that makes sense I see

devmario commented on anime.js:
```diff
@@ -590,22 +590,33 @@
-        if(is.undef(callback) || !is.func(callback)) return this;
+        var s = this.settings;
+        if(!is.func(callback)) {
+            s.update = undefined;
```
If you setup not valid argument, callback is disapear.

SaulDoesCode replied:
if you return this before ``s.complete = callback;`` then that wouldn't happen.

devmario replied:
Is this not good?
Any idea?

SaulDoesCode replied:
I like that you can then disable the callback from with in it like in your example
```javascript
anim.play()
.complete(anim => {
   anim.play().complete(); // This code is complete callback disable
});
```
the only problem I see is ambiguity why not add a disabler to the anim object or do it manually
```javascript
anim.play()
.complete(anim => {
     anim.settings.complete = undefined; 
});
```

SaulDoesCode replied:
```javascript
    anim.complete = function(callback) {
        var s = anim.settings;
        if(!is.func(callback)) return anim;
        else s.complete = callback;
        if(callback === true) s.complete = undefined;
        return anim;
    }
```

devmario replied:
How about this?
```js
  anim.complete = function(callback) {
        var s = anim.settings;
        if(is.func(callback)) s.complete = callback;
        else if(is.null(callback)) s.complete = undefined;
        return anim;
    }
```

SaulDoesCode replied:
no actually I like your idea 
```javascript
// this is perfect
anim.play().complete(anim => {
         anim.play().complete(); // callback disable
   });
```

SaulDoesCode replied:
```javascript
    // anything that is not a function disables it
    anim.complete = function(callback) {
        anim.settings.complete = is.func(callback) ? callback : undefined;
        return anim;
    }
```

devmario replied:
@SaulDoesCode 
Impletment just now

SaulDoesCode replied:
if you wanna cheat and make it really tiny to implement maybe try this
it's the same function for all three essentially so make a factory / maker function / thing
```javascript
       var callbacks = function(type) {
      return function(callback) {
        anim.settings[type] = is.func(callback) ? callback : undefined;
        return anim;
      }
    }
    anim.begin = callbacks('begin');
    anim.update = callbacks('update');
    anim.complete = callbacks('complete');
```

devmario replied:
We think the same.
I'll like that

devmario replied:
```js
var callbacks = function(type) {
      return function(callback) {
        anim.settings[type] = is.func(callback) ? callback : undefined;
        return anim;
      }
    };

    anim.begin = callbacks('begin');
    anim.update = callbacks('update');
    anim.complete = callbacks('complete');
```

Like this?

SaulDoesCode replied:
the es6 version looks so weird though, but it works :laughing: 
```javascript
           const callbacks = type => callback => {
              anim.settings[type] = is.func(callback) ? callback : undefined;
              return anim;
            }
```

devmario replied:
There seems fine to me.
Cool

luknei commented:
Loosk like promise support wont be available?

SaulDoesCode commented:
perhaps it wont be necessary, the chainable methods work very similar and it's much more compatible with older platforms than using promises.

luknei commented:
Could you provide example how would you sequence animations. Let's say you wait for two components animations to end and then run third animation?

SaulDoesCode commented:
Check the readmes of either @devmario's fork of version v1.1 or my own they have demo's. I think @devmario made a jsfiddle demo somewhere I'll try and find the link for you

SaulDoesCode commented:
https://jsfiddle.net/SaulDoesCode/0wkyLo6k/

luknei commented:
This is not what I was looking for, since there is no animations sequencing.
I'm talking about example like this:
Animation A: starts immediately, runs for .5s
Animation B: starts immediately, runs for 1s
Animation C: starts after A & B is completed, runs for 1s

SaulDoesCode commented:
I think that might be somewhat beyond the expectations of this library perhaps just make those animations
and use the .complete , .begin , .update  methods as time based indicators of when to start or stop other animations.
What do you propose the best hypothetical syntax to achieve this would be?

lukebatchelor commented:
I assume @luknei wants to be able to use things like, Promise.all and Promise.any?

```
Promise.all([anim1, anim2, anim3]).then(() => { 
  // anime(...)
});
```

devmario commented:
Thx @lukebatchelor @luknei @SaulDoesCode for good idea.
We need implement Promise.all and Promise.any style.
How about for promise naming ".completed, began, updated"?
#4

luknei commented:
Well I tried observables also wrote my own timeline for sequencing animations. However, from my experience, the best solution so far was to use promises. Since it is one of the language fundamentals now and it is quite expressive - it is easy to read, write and reason about.
Example:

`
Promise.all([
    flyInLeft('.blue-box'),
    flyInRight('.green-box')
]).then(([animBlueBox, animGreenBox]) => {
    return fadeIn('.black-box');
}).then(animBlackBox => {
    return fadeOut(['.blue-box', '.green-box', '.black-box'])
}).then(() => {
    console.log('animations sequence completed!');
});
`

Even though it might be quite a complex animations sequence, promises makes it easy to understand what is going on.

SaulDoesCode commented:
Hang on I'm working on a little hack that might provide an alternative that is exactly like that but won't need promises.

luknei commented:
@devmario i was thinking about the same option. Adding began & completed properties to animation instance. If we would make them with a getters it would be possible to create Promises lazily. In that way it would be completely opt-in feature, if you do not use those properties (began, completed), even browsers without Promise support would not break;

devmario commented:
I implemented promise. @SaulDoesCode @luknei @lukebatchelor 

Check my fiddle
https://jsfiddle.net/devmario/r95z5f6e/1/

How about this style?
```js
Promise.all([anim1.completed(), anim2.completed(), anim3.completed()]).then(function() {
    console.log("all completed");
    anim1.began().then(function() {
        console.log("began");
    });
    anim1.play().completed().then(function() {
       console.log("completed");
    });
});
```

SaulDoesCode commented:
@devmario your implementation looks really good, I like that you kept the chainable style as well.

devmario commented:
May be write this example with promise information at README.md
```js
Promise.all([anim1.completed(), anim2.completed(), anim3.completed()]).then(function() {
    console.log("all completed");
    anim1.play().began().then(function() {
        console.log("began");
    });
    anim1.updated().then(function() {
        console.log("updated callback only once in promise");
    });
    anim1.completed().then(function() {
       console.log("completed");
    });
});
```

SaulDoesCode commented:
maybe we could add some wrappers to make it nicer to work with
```javascript
    anime.alldone([anim1,anim2,anim3]).then(func);
    anime.allbegan([anim1,anim2,anim3]).then(func);

   // possible implementation
   animation.alldone = function(anims) {
       if(is.object(anims)) anims = toArray(arguments);
       anims = anims.map(function(anim) {
            return anim.completed();
       });
       return Promise.all(anims);
   }
```

SaulDoesCode commented:
What is wrong with polyfills?
```javascript
 Promise.toString().indexOf("[native code]") !== -1
```

devmario commented:
@SaulDoesCode 
Perhaps because of the following differences:

```js
var p = anim1.completed();
Promise.all([p]).then(function() {
  console.log('call');
});
Promise.all([p]).then(function() {
  console.log('call one more time');
});
```

```js
Promise.all([anim1.completed()]).then(function() {
  console.log('don't call');
});
Promise.all([anim1.completed()]).then(function() {
  console.log('call');
});
```
If using the 'alldone' and 'allbegan', I think need to change the 'alldone'  argument to be a promise array type.
We have to be able to give a variety of ways to use our anime. ;)

luknei commented:
@devmario Really nice! However, I'd like to make suggestion to make began and completed as properties instead of functions and memoize the getters of those properties. And we need to think what should happen in case  of restart function is called. I would suggest to create new animation instance with the same parameters. What are your thought guys?
@SaulDoesCode I do not really have a strong opinion on the wrappers idea :|

devmario commented:
@luknei
Have you ever able to show a good example to me like before?

luknei commented:
@devmario 

````

var memoize = function(fn) {
    var result;
    
    return function() {
      if (result) return result;
      
      result = fn.apply(null, arguments);
      
      return result;
    }
  }
Object.defineProperty(anim, 'began', {
  enumerable: true,
  configurable: false,
  get: memoize(function () {
   return new Promise(function () {});
  })
  //or  in your case: "get: memoize(promise('began'))"
});

...
```

devmario commented:
Thx @luknei
Your code that makes me two good ideas.
- Our anime properties must be readable!
- We need memorization for promise properties.

How about this?
```js
anim.completed(); //new Promise always
anim.completed(key); //new Promise at another key
```
Or
```js
anim.completed(); //new Promise only once
anim.completed(key); //new Promise at another key
```

SaulDoesCode commented:
I like the concept of getters for promises it just looks nice.
```javascript
        function promise(type) {
                if (typeof Promise !== "undefined") {
                  Object.defineProperty(anim,type,{
                    get() {
                        return new Promise(function(resolve) {
                             return (anim[type + '_resolve'] = resolve)
                        });
                    },
                  });
                } else console.warn("Your browser doesn't support promises.");
            }
           
            promise('began');
            promise('updated');
            promise('completed');


           // then you can do 
          anim.complete.then(func)
          // or
          Promise.all([anim.complete,anim2.complete,anim2.complete]).then(func);
```

luknei commented:
@devmario thanks (: I did not get what is that key you are passing to completed?

```
anim.completed(key);
```

SaulDoesCode commented:
Yeah what does the `` anim.completed(key); `` key do?

devmario commented:
Thx @SaulDoesCode 
I will apply your code. ;)

@luknei @luknei 
I thought strange about **anim.completed(key)**. I'm tired. I'll sleep. I was awake for 24 hours.

SaulDoesCode commented:
Yeah you did a lot of coding, get some rest

ivannn1234 commented on README.md:
```diff
@@ -249,23 +249,65 @@ anime({
+myAnimation.began().then(function(anim) {
+  console.log("Resolved began promise!");
+});
+
```
(anim.settings.update

devmario replied:
Promise update also will be needed.
Because detected will be needed once it changes to play when pause.
For example
```js
myAnimation.updated().then(function(anim) {
   console.log("call only once!");
});
```

SaulDoesCode replied:
What about implementing an event system I found that to have been the most simple solution on my es6 fork; it aslo supports promises but uses the even system underneath
```javascript
  var myAnimation = anime({
    targets: 'div',
    translateX: 100,
    autoplay: false
  });

  myAnimation.once('begin',anim => {
    console.log("Began!"); // Called the animation began.
  });

  myAnimation.once('complete',anim => {
      console.log("Completed!"); // Called the animation ended.
  });

  let listener = myAnimation.on('update', anim => {
    console.log("Updated!"); // Called the animation updated.
  });
  listener.off();
  // later
  listener.on();
  // all of the above events may be accessed via the .on or .once methods
  myAnimation.once('pause',(anim,listener) => {
    console.log("the animation was paused!"); // called on pause
    setTimeout(anim.play, 1500);
  });

  myAnimation.play(); // Manually play the animation

```

devmario replied:
@SaulDoesCode  
I think good things that event system familiar to us as another javascript library.

SaulDoesCode replied:
the event system is not an external dependency it's all still self contained, just tested some demo's now seems solid and the performance is good.
I quite like the node.js style `` .on `` and I think people are familiar with it like you say
maybe people would like it in anime. 
The other big problem v1.1 still faces is having multiple callbacks or "listeners".
and I think it needs to be solved. 
```javascript 
  // this might not be enough for users who want to have multiple callbacks
  if(is.func(anim.settings.callback)) anim.settings.callback(x,y,z);
```

devmario replied:
Maybe we just got even better be able to remove a callback
If we use event system, we need to be clean namespace for callback, event name, promise property name.
You have an idea?

juliangarnier commented:
Not sure to understand the benefits of using promises in this case, and why we should have both:

```
    myAnimation.complete(function(anim) {
      console.log('completed ' + anim);
    });

    myAnimation.completed().then(function() {
      console.log('completed ' + anim);
    });

```

Am I missing something?

SaulDoesCode commented:
I agree, I think the promises are maybe a bit much

devmario commented:
Maybe we just got even better be able to remove a callback.
I'll remove callback.

SaulDoesCode commented:
There's still the problem of having multiple "listeners" on animations, my event system based solution solves that problem also allows capturing more types of events at different stages like 
``anim.on('pause',callback)``

devmario commented:
And 'anim.once'

SaulDoesCode commented:
var myAnimation = anime({
    targets: 'div',
    translateX: 100,
    autoplay: false
  });

  myAnimation.once('begin',anim => {
    console.log("Began!"); // Called the animation began.
  });

  myAnimation.once('complete',anim => {
      console.log("Completed!"); // Called the animation ended.
  });

  let listener = myAnimation.on('update', anim => {
    console.log("Updated!"); // Called the animation updated.
  });
  listener.off();
  // later
  listener.on();

  // all of the above events may be accessed via the .on or .once methods
  myAnimation.once('pause',(anim,listener) => {
    console.log("the animation was paused!"); // called on pause
    setTimeout(anim.play, 1500);
  });

  myAnimation.play(); // Manually play the animation

devmario commented:
About listener
```Js
  let listener = myAnimation.on('update', anim => {
    console.log("Updated!"); // Called the animation updated.
  });
  listener.off();
  // later
  listener.on();
```
How about this?

```Js
  myAnimation.on('updated', anim => {
    console.log("Updated!"); // Called the animation updated.
  }).off('updated'); // or myAnimation.off() in this case all event not occur
  // later
  myAnimation.on('updated'); // or myAnimation.on() in this case all event occur at timing of each
```

SaulDoesCode commented:
not sure about 
```javascript
  myAnimation.on('updated', anim => {
    console.log("Updated!"); // Called the animation updated.
  }).off('updated'); // or myAnimation.off() in this case all event not occur
  // later
  myAnimation.on('updated'); // or myAnimation.on() in this case all event occur at timing of each
```
because a listener is anonymouse and not attached to the anim object at all.
that means that if you don't assign it to something it is lost
I'm still working out a couple of kinks with it but it's quite performant
```javascript
  let listener =  myAnimation.on('update', anim => {
    console.log("Updated!"); // Called the animation updated.
  }).off();
 listener.on();
```

devmario commented:
@juliangarnier 
About your question
For example
```js
 myAnimation.complete(function(anim) {
      console.log('completed ' + anim);
 });

myAnimation.completed.then(function(anim) {
      console.log('completed ' + anim); // Call only once
});

// with
Promise.all([myAnimation.completed, myAnimation2.completed]).then(function(anims) {
       // Called once at completed twice 
});
```

SaulDoesCode commented:
one example of abstracting around event listeners
```javascript
 animation.alldone = function () {
            const anims = flattenArr(arguments);
            return new Promise(pass => {
                let i = 1;
                anims.forEach(anim => {
                    anim.once('complete', () => {
                        if (i++ === anims.length) pass(anims);
                    });
                });
            });
        }

     let anims = Array.from(document.querySelectorAll('.button-wrapper > section')).map(el => {
            return anime({
                targets: el,
                translateY : -30,
                duration:anime.random(100,2000),
            });
          });

        anime.alldone(anims).then(anims => {
          console.log('all done! :D',anims);
        });

```

devmario commented:
How about my todo list
- event implementation
- callback remove

And about namespace
- promise(property)
 - completed
 - began
 - updated
- event name(Just equal promise property name)
 - began
 - completed
 - updated

SaulDoesCode commented:
```javascript
    if (typeof Promise != "undefined") {
                function promise(type) {
                    Object.defineProperty(anim, type, {
                        get() {
                            return new Promise(pass => {
                                anim.once(type, pass);
                            });
                        },
                    });
                }

                promise('completed');
                promise('began');
               // not sure about update as it needs to be called more than once
            } else console.warn("anime : Your browser doesn't support promises.");

```

devmario commented:
@SaulDoesCode how about this?
```js
if (typeof Promise != "undefined") {
       anim.promise = function(type) {
              ...
       }
} else console.warn("anime : Your browser doesn't support promises.");
if(is.func(anim.promise)) {
       promise("completed");
       ...
}
```

SaulDoesCode commented:
Yeah that will work, I see no problem with that. 
I apologize for all the spam on here I am sorry. I want to show you how the event system is implemented

```javascript
function eventemitter(obj) {
        if(!obj) obj = {};
        let options = {
            evtlisteners: new Set,
            __stop: false,
            on(type, func) {
                if (!is.func(func)) throw new TypeError(`.on(${type},func) : func is not a function`);
                func.etype = type;
                func.ehandle = {
                    on() {
                        func.etype = type;
                        options.evtlisteners.add(func);
                        return func.ehandle;
                    },
                    off: () => {
                      options.off(func);
                      return func.ehandle;
                    },
                };
                options.evtlisteners.add(func);
                return func.ehandle;
            },
            once(type, func) {
                if (is.func(func)) {
                    function funcwrapper() {
                        func.apply(obj || this, arguments);
                        options.off(funcwrapper);
                    }
                    return options.on(type, funcwrapper);
                } else throw TypeError('eventemitter: .once needs a function');
            },
            off(func) {
                if (options.evtlisteners.has(func)) options.evtlisteners.delete(func);
                return options;
            },
            emit(type) {
                if (!options.stop) {
                    let args = Array.prototype.slice.call(arguments, 1);
                    options.evtlisteners.forEach(ln => {
                        if (ln.etype == type && !options.__stop) ln.apply(this, args.concat(ln.ehandle));
                    });
                }
                return options;
            },
            stopall(stop) {
                options.__stop = is.bool(stop) ? stop : true;
            },
        };
        Object.keys(options).forEach(key => {
            Object.defineProperty(obj, key, {
                value: options[key],
                enumerable: false,
                writable: false,
            })
        });
        return obj;
    }
```

juliangarnier commented:
Let's not overthink it.

I want to keep the anime.js lightweight and simple as possible.

I think adding begin() / update() / complete() methods is a good idea because it helps readability.
And promises are interesting too.

But you're suggesting 4 ways to do **roughly** the same thing:

```
    var myAnimation = anime({
      targets: 'div',
      translateX: 100,
      complete: function(anim) {
        console.log('completed ' + anim);
      }
    });

    myAnimation.complete(function(anim) {
      console.log('completed ' + anim);
    });

    myAnimation.completed().then(function(anim) {
      console.log('completed ' + anim);
    });

    myAnimation.on('complete', function(anim) {
      console.log('completed ' + anim);
    });
```

I can't accept this kind of overlap in the API.

SaulDoesCode commented:
Then let us settle on one and standardize it

devmario commented:
@juliangarnier I agree
Then promise and event system case, What do you think about.
Should support promise and event or not?
Or would fine if we clean the namespace?

juliangarnier commented:
Why not simply adding promises to the old methods?

like this:
```
    myAnimation.complete(function(anim) {
      console.log('completed ' + anim);
    }).then(function(anim) {
      console.log('completed once' + anim);
    });
```

That way we keep the old behaviours unchanged, while adding promises.

What do you think?

SaulDoesCode commented:
I think that adds a bit of ambiguity in the mix returning a promise while optionally executing a callback above might be a bit confusing.

devmario commented:
```js
// If we support like this
myAnimation.complete(function(anim) {
      console.log('completed ' + anim);
}).then(function(anim) {
      console.log('completed once' + anim);
}); 

// don't support like this 
myAnimation.complete(function(anim) {
      console.log('completed ' + anim);
})
.begin(); 
```

But if we use apply method
```js
anim.apply(new Promise());
```
Is it available?

SaulDoesCode commented:
with the event system we can support the in parameter functions simply like so maintaining compatibility 
```javascript
anime({
  targets: 'div',
  translateX: '3rem',
  width: '100',
  update:function(anim) {
     // do stuff
  },
  begin: function(anim) {
     // do stuff
  },
  complete:function(anim) {
     // do stuff
  }
});
   
  // so internally it gets handled like this
  ['complete', 'begin', 'update'].forEach(function(type) {
           if (is.func(anim.settings[type])) anim[type == 'update' ? 'on':'once'](type, anim.settings[type]);
   });
```

juliangarnier commented:
@SaulDoesCode I like the idea, I'm interested to see a clean PR for the event system.
I'm just a little bit afraid on the weight cost. I really want to keep the library under 10KB minified.

SaulDoesCode commented:
I'll work at it and cut down as much bulk as possible.

SaulDoesCode commented:
I'll start by forking a clean v1.1, may I ask that you perhaps make a new experimental branch and don't touch it for a while so I can rewrite all the necessary bits without too much continuous re-integration

juliangarnier commented:
@SaulDoesCode here: https://github.com/juliangarnier/anime/tree/event-system

@devmario I think we can close this PR now, but you can make a new one just for the build system

SaulDoesCode commented:
thanks
